### PR TITLE
2chip gbs

### PR DESCRIPTION
**Pull request by ziofil**
### **User description**
**Context:**
a few chores

**Description of the Change:**
* made some transpose operations compatible with batch dimensions
* Disabled `parallel=True` for `vanilla_batch_numba` in the JAX backend
* updated CHANGELOG

___

### **PR Type**
Enhancement

___

### **Description**
- Refactored JAX backend matmul for better batch dimension support

- Improved transpose operations to work with batch dimensions  

- Added NUMBA_PARALLEL setting for threading control

- Updated CHANGELOG with recent enhancements and fixes

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["JAX Backend"] --> B["Sequential matmul"]
  C["Backend Manager"] --> D["swapaxes operations"]
  E["Numba Settings"] --> F["NUMBA_PARALLEL control"]
  G["Parameter Update"] --> H["swapaxes transpose"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>backend_jax.py</strong><dd><code>Refactor matmul to use sequential multiplication</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

mrmustard/math/backend_jax.py

<ul><li>Replaced <code>jnp.linalg.multi_dot</code> with iterative <code>jnp.matmul</code> calls<br> <li> Improved control over matrix multiplication operations</ul>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/642/files#diff-277d6b9d9077473e0fca45c03a82a1503f283442e0b7503da968e5330ec3c242">+4/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>backend_manager.py</strong><dd><code>Update gradient operations to use swapaxes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

mrmustard/math/backend_manager.py

<ul><li>Replaced <code>transpose</code> with <code>swapaxes(-1, -2)</code> in gradient calculations<br> <li> Updated both <code>euclidean_to_symplectic</code> and <code>euclidean_to_unitary</code> methods</ul>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/642/files#diff-272a316cdf15707ddb7384f13258135f593fc560e21dc19cf17c16cfb5d0aa1c">+4/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>parameter_update.py</strong><dd><code>Update transpose to swapaxes in parameter updates</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

mrmustard/training/parameter_update.py

- Replaced `transpose` with `swapaxes(-1, -2)` for consistency

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/642/files#diff-9024523f024f0b3c9ca5effad5560daf9d84a5e44b95b712e36d305caeb7bcf7">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>batch.py</strong><dd><code>Use configurable NUMBA_PARALLEL setting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

mrmustard/math/lattice/strategies/vanilla/batch.py

<ul><li>Added import for <code>settings</code> module<br> <li> Changed parallel parameter to use <code>settings.NUMBA_PARALLEL</code></ul>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/642/files#diff-5f357db285994a7c36f9d65ff226664e3e1d8313fbbea13a5e21eb93af0cefcf">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>settings.py</strong><dd><code>Add NUMBA_PARALLEL configuration setting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

mrmustard/utils/settings.py

<ul><li>Added <code>NUMBA_PARALLEL</code> boolean setting with default <code>True</code><br> <li> Includes documentation about JAX backend compatibility</ul>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/642/files#diff-eaa5467c3f570ae6efabf33215bd7a36eb842dcec72aa1799c3bfb43b6e9f6b7">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Comprehensive changelog update with recent changes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/CHANGELOG.md

<ul><li>Added extensive changelog entries for recent enhancements<br> <li> Documented bug fixes, documentation improvements, and tooling updates<br> <li> Added entries for Riemannian optimizers, Fock improvements, and <br>backend changes</ul>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/642/files#diff-2a22f598a15a364508bc9a475e6ec9df31958a753e9401fabaac8df4db5bd853">+79/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: mrmustard/math/backend_jax.py [hunk 1]
```diff
@@ -299,7 +299,10 @@ def make_complex(self, real: jnp.ndarray, imag: jnp.ndarray) -> jnp.ndarray:
 
     @jax.jit
     def matmul(self, *matrices: jnp.ndarray) -> jnp.ndarray:
-        return jnp.linalg.multi_dot(matrices)
+        result = matrices[0]
+        for m in matrices[1:]:
+            result = jnp.matmul(result, m)
```
Comment (heltluke) [inline mrmustard/math/backend_jax.py]: What was wrong with `multi_dot`?

**Review by heltluke — DISMISSED** [review]

File: mrmustard/math/backend_jax.py (hunk 1 above)
Comment (apchytr) [reply mrmustard/math/backend_jax.py]: Updated in [635](https://github.com/XanaduAI/MrMustard/pull/635/files#diff-277d6b9d9077473e0fca45c03a82a1503f283442e0b7503da968e5330ec3c242R290)

Comment (apchytr) [reply mrmustard/math/backend_jax.py]: But basically its to handle batch dims

Comment (heltluke) [reply mrmustard/math/backend_jax.py]: Ah, it's on account of batch dimensions. Looks like it'll stay part of a `try-except` when merging `develop` anyway.

File: mrmustard/math/lattice/strategies/vanilla/batch.py [hunk 2]
```diff
@@ -17,14 +17,15 @@
...
+from mrmustard import settings
 from mrmustard.utils.typing import ComplexTensor
 
 from .core import stable_numba, vanilla_numba
 
 # ruff: noqa: RUF005
 
 
-@njit(cache=True, parallel=True)
+@njit(cache=True, parallel=settings.NUMBA_PARALLEL)
```
Comment (apchytr) [inline mrmustard/math/lattice/strategies/vanilla/batch.py]: Does this have the previously mentioned issue where because its a decorator it won't pick up changes to the settings?

File: mrmustard/utils/settings.py [hunk 3]
```diff
@@ -101,6 +101,10 @@ def __init__(self):
         self.STABLE_FOCK_CONVERSION: bool = False
         r"""Whether to use the ``stable`` function when computing Fock amplitudes (more stable, but slower). Default is ``False``."""
 
+        self.NUMBA_PARALLEL: bool = True
```
Comment (apchytr) [inline mrmustard/utils/settings.py]: This seems somewhat weird to me as a setting. Particularly, because this is backend dependent and therefore we shouldn't expect users to be able to change this. Perhaps the backend manager should be handling this

File: .github/CHANGELOG.md [hunk 4]
```diff
@@ -19,6 +19,49 @@
 * Added vjps to the JAX backend. Added ``numba`` based Fock lattice strategies to `SqueezedVacuum` and `Sgate`.
 [(#636)](https://github.com/XanaduAI/MrMustard/pull/636)
 
+* Riemannian updates to `OptimizerJax` for improved convergence.
```
Comment (apchytr) [inline .github/CHANGELOG.md]: Can we reorder these to be in order of merging?

**Review by apchytr — CHANGES_REQUESTED** [review]
Just a few comments!

File: .github/CHANGELOG.md (hunk 4 above)
Comment (ziofil) [reply .github/CHANGELOG.md]: why? (it's going to take a while)

File: mrmustard/utils/settings.py (hunk 3 above)
Comment (ziofil) [reply mrmustard/utils/settings.py]: Yes, I thought about that, but then do you kill a user's kernel when they change backend? Or do you instruct them to do so? I thought in the end it was much simpler as a setting with documented behaviour, than to come up with a mechanism that detects the environment and (interactively?) allows the system to kill the kernel.

File: .github/CHANGELOG.md [hunk 5]
```diff
@@ -19,6 +19,49 @@
...
+[(#571)](https://github.com/XanaduAI/MrMustard/pull/571)
+
+* Reduced memory overhead by replacing `np.array` with `np.asarray` where appropriate.
+[(#600)](https://github.com/XanaduAI/MrMustard/pull/600)
+
+* Refactored JAX backend `matmul` and associated gradient operations for clarity and control.
+
+* Improved batch support in transpose operations.
+
+* Improved control of Numba parallel execution in `vanilla_batch_numba` via `NUMBA_PARALLEL` setting.
```
Comment (ziofil) [inline .github/CHANGELOG.md]:
```suggestion
* Refactored JAX backend `matmul` and associated gradient operations for clarity and control.
* Improved batch support in transpose operations.
* Improved control of Numba parallel execution in `vanilla_batch_numba` via `NUMBA_PARALLEL` setting.
[(#642)](https://github.com/XanaduAI/MrMustard/pull/642)
```

File: mrmustard/utils/settings.py (hunk 3 above)
Comment (apchytr) [reply mrmustard/utils/settings.py]: Pushed a change where math backend now handles it so no longer need a setting 👍

**Review by apchytr — APPROVED** [review]
